File: Loenn/entities/LaserEmitter.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local LaserEmitter = {}

LaserEmitter.name = "XaphanHelper/LaserEmitter"
LaserEmitter.depth = -2000
LaserEmitter.canResize = {false, false}
LaserEmitter.fieldInformation = {
    side = {
        options = {"Left", "Right", "Top", "Bottom"},
        editable = false
    },
    type = {
        options = {"Kill", "Must Dash", "No Dash"},
        editable = false
    }
}
LaserEmitter.placements = {
    name = "LaserEmitter",
    data = {
        flag = "",
        side = "Right",
        type = "Kill",
        directory = "objects/XaphanHelper/LaserEmitter",
        inverted = false,
        base = true,
        noBeam = false
    }
}

function LaserEmitter.sprite(room, entity)
    local side = entity.side or "Left"
    local directory = entity.directory or "objects/XaphanHelper/LaserEmitter"
    local base = entity.base

    local sprites = {}

    local sprite = drawableSprite.fromTexture(directory .. "/idle00", entity)
    local baseSprite = drawableSprite.fromTexture(directory .. "/baseActive" .. side .. "00", entity)
    
    sprite:addPosition(4, 4)
   
    if side == "Bottom" then
        sprite.rotation = math.pi
        baseSprite:addPosition(4, -4)
    elseif side == "Top" then
        sprite.rotation = 0
        baseSprite:addPosition(4, 12)
    elseif side == "Left" then
        sprite.rotation = -math.pi / 2
        baseSprite:addPosition(12, 4)
    elseif side == "Right" then
        sprite.rotation = math.pi / 2
        baseSprite:addPosition(-4, 4)
    end

    sprite:setColor("White")

    if sprite then
        table.insert(sprites, sprite)
    end
    if base and baseSprite then
        table.insert(sprites, baseSprite)
    end

    return sprites
end

function LaserEmitter.selection(room, entity)
    local side = entity.side or "Left"
    local base = entity.base
    local xAdjust = 0
    local width = 0
    local yAdjust = 0
    local height = 0
    if side == "Bottom" then
        width = 8
        height = 2
        if base then
            height = 10
            yAdjust = -8
        end
    elseif side == "Top" then
        width = 8
        height += 2
        yAdjust = 6
        if base then
            height = 10
        end
    elseif side == "Left" then
        height = 8
        width = 2
        xAdjust = 6
        if base then
            width = 10
        end
    elseif side == "Right" then
        height = 8
        width = 2
        if base then
            width = 10
            xAdjust = -8
        end
    end

    return utils.rectangle(entity.x + xAdjust, entity.y + yAdjust, width, height)
end

return LaserEmitter

File: Ahorn/entities/LaserEmitter.jl
module XaphanHelperLaserEmitter

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/LaserEmitter" LaserEmitter(x::Integer, y::Integer, flag::String="", side::String="Right", type::String="Kill", directory::String="objects/XaphanHelper/LaserEmitter", inverted::Bool=false, base::Bool=true, noBeam::Bool=false)

const placements = Ahorn.PlacementDict(
    "Laser Emitter (Xaphan Helper)" => Ahorn.EntityPlacement(
        LaserEmitter,
        "point",
        ),
    )


Ahorn.resizable(entity::LaserEmitter) = false, false

function Ahorn.selection(entity::LaserEmitter)
    x, y = Ahorn.position(entity)
    side = get(entity.data, "side", "Left")
    base = get(entity.data, "base", true)
    if side == "Left"
        return Ahorn.Rectangle(x, y, 16, 8)
    end
    if side == "Right"
        return Ahorn.Rectangle(x - 8, y, 16, 8)
    end
    if side == "Top"
        return Ahorn.Rectangle(x, y, 8, 16)
    end
    if side == "Bottom"
        return Ahorn.Rectangle(x, y - 8, 8, 16)
    end
end

Ahorn.editingOptions(entity::LaserEmitter) = Dict{String, Any}(
    "side" => String["Left", "Right", "Top", "Bottom"],
    "type" => String["Kill", "Must Dash", "No Dash"]
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LaserEmitter, room::Maple.Room)
    side = get(entity.data, "side", "Right")
    directory = get(entity.data, "directory", "objects/XaphanHelper/LaserEmitter")
    base = get(entity.data, "base", true)

    if side == "Left"
        Ahorn.drawSprite(ctx, "$(directory)/idle00.png", 4, 12, rot=-pi / 2)
        if base
            Ahorn.drawSprite(ctx, "$(directory)/baseActive$(side)00.png", 12, 4)
        end
    end
    if side == "Right"
        Ahorn.drawSprite(ctx, "$(directory)/idle00.png", 12, 4, rot=pi / 2)
        if base
            Ahorn.drawSprite(ctx, "$(directory)/baseActive$(side)00.png", -4, 4)
        end
    end
    if side == "Bottom"
        Ahorn.drawSprite(ctx, "$(directory)/idle00.png", 12, 12, rot=pi)
        if base
            Ahorn.drawSprite(ctx, "$(directory)/baseActive$(side)00.png", 4, -4)
        end
    end
    if side == "Top"
        Ahorn.drawSprite(ctx, "$(directory)/idle00.png", 4, 4)
        if base
            Ahorn.drawSprite(ctx, "$(directory)/baseActive$(side)00.png", 4, 12)
        end
    end
end

end